Added support for Ampq v1.0

AndreSteenbergen opened the pull request.

AndreSteenbergen committed:
Added support for Ampq v1.0

Aaronontheweb commented:
@AndreSteenbergen so is this going to publish as its own independent package?

AndreSteenbergen commented:
I think it should, as amqp 0.9 (RabbitMq and the likes) and amqp 1.0 are incompatible. With RabbitMQ being quite popular, I don't think it is wise to remove 0.9 support.

AndreSteenbergen commented:
BTW this pull request is just a first start to support AMQP 1.0. I hope good feedback can be given to the code, I hope this contributes something .

AndreSteenbergen commented:
This pull request was made to support this: https://github.com/AkkaNetContrib/Alpakka/issues/37

Aaronontheweb commented:
Ah, got it.

CC @bobanco - any thoughts on this?

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSinkStage.cs:
```diff
@@ -0,0 +1,71 @@
+    public sealed class AmqpSinkStage<T> : GraphStageWithMaterializedValue<SinkShape<T>, Task>
+    {
+        public readonly Inlet<T> In = new Inlet<T>("AmqpSink.in");
+        public override SinkShape<T> Shape => new SinkShape<T>(In);
```
It's most likely that the getter is called multiple times during materialization, i.e. wenn calling `.Run`, so  this should be set once in the constructor.

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSinkStage.cs:
```diff
@@ -0,0 +1,71 @@
+                    try
+                    {
+                        var elem = Grab(stage.In);
+                        sender.Send(new Message(amqpSinkStage.AmpqSourceSettings.GetBytes(elem)));
```
Is this a blocking call ?

AndreSteenbergen replied:
There is an async version. but should I await it, or just continue?

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSinkStage.cs:
```diff
@@ -0,0 +1,71 @@
+                        Pull(stage.In);
+                    } catch (Exception e)
+                    {
+                        Debugger.Break();
```
Do you forgot to remove it or because it's still WIP ?

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSinkStage.cs:
```diff
@@ -0,0 +1,71 @@
+
+            public override void PostStop()
+            {
+                promise.TrySetException(new ApplicationException("stage stopped unexpectedly"));
```
If the stream fails `onUpstreamFailure` is called, so I'm not sure why we should try to "throw" another exception from her.

vasily-kirichenko replied:
Anyways, why `ApplicationException`? I believe MS encouraged to not use it many years ago.

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+            AmqpSourceSettings = amqpSourceSettings;
+        }
+
+        public override SourceShape<T> Shape => new SourceShape<T>(Out);
```
As above, set once via constructor

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+                decider = attributes.GetDeciderOrDefault();
+
+                SetHandler(outlet, () => {
+                    var message = receiver.Receive();
```
Is there no async version that could be used instead ?

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+
+                SetHandler(outlet, () => {
+                    var message = receiver.Receive();
+                    if (message != null)
```
What if `message` is `null` ? You probably need to call `Receive` again ?!

vasily-kirichenko replied:
Can the receiver return `null` at all? If so, what it means?

AndreSteenbergen replied:
I don't know, I thought this OnPull was called by Akka every x times per second? I just read the comment below. Yes you are correct, I would need to go return to Receive

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+                            {
+                                FailStage(e);
+                            }
+                        }
```
What if `ampqSourceSettings.Convert(message)` fails and decider returns `Resume` or `Restart` ? 
You haven't pushed a message downstream and therefore the `OnPull` callback isn't called again because downstream still waits for an element, but you aren't trying to receive another message from the queue, hence your source would idle forever ;-)

AndreSteenbergen replied:
I think I need to change this, more like the other amqp version. Will do now.

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+                        try
+                        {
+                            Push(outlet, ampqSourceSettings.Convert(message));
+                            receiver.Accept(message);
```
Is it a good idea to immediately accept the message ? 
What if I would want to accept it at the end of the stream because processing could fail later in the chain and if so I want to try to process it a second time?

AndreSteenbergen replied:
How would that work? Sending a reference to the receiverlink with the message.

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+                        }
+                        catch (Exception e)
+                        {
+                            receiver.Reject(message, new Error(new Symbol(e.Message)));
```
Not sure about the semantics of `Reject` but would it maybe make sense to sync it with the result of the decider ?  
E.g. if decider returns `Resume` or `Restart` one maybe doesn't want to reject the message but instead try it another time ?

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/Dsl/AmpqSink.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public static class AmpqSink<T>
+    {
+        public static Sink<T, Task> Create(IAmpqSinkSettings<T> sourceSettings)
```
I would move the `T` into the method definition, in theory you should be able to call `AmpqSink.Create(settings)` because the `T` could be inferred from the settings.

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/Dsl/AmpqSource.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public static class AmpqSource<T>
+    {
+        public static Source<T, NotUsed> Create(IAmqpSourceSettings<T> sourceSettings)
```
As above, move the `T`

marcpiechura commented on Amqp.V1/Akka.Streams.Amqp.V1/NamedQueueSourceSettings.cs:
```diff
@@ -0,0 +1,35 @@
+            this.serializer = serializer;
+        }
+
+        public T Convert(Message message) {
```
Move `{` into the next line ;-)

marcpiechura commented:
@AndreSteenbergen  left you a few remarks,. One more general question, aren't there any async methods for the queue, e.g. `ReceiveAsync`, `AcceptAsync` ... ?

vasily-kirichenko commented on Amqp.V1/Akka.Streams.Amqp.V1.Tests/AmqpConnectorsTest.cs:
```diff
@@ -0,0 +1,55 @@
+        public AmqpConnectorsTest()
+        {
+            materializer = ActorMaterializer.Create(Sys);
+            var serialization = Sys.Serialization;
```
Why this variable?

vasily-kirichenko commented on Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs:
```diff
@@ -0,0 +1,68 @@
+            public override void PostStop()
+            {
+                base.PostStop();
+                receiver.Close();
```
Are you sure we should call the base method first?

AndreSteenbergen replied:
thx, nope

vasily-kirichenko commented on Amqp.V1/Akka.Streams.Amqp.V1/NamedQueueSinkSettings.cs:
```diff
@@ -0,0 +1,34 @@
+        private readonly Session session;
+        private readonly string linkName;
+        private readonly string queueName;
+        private readonly Serializer serializer;
```
Why not CamelCase for all these prop names? Is it something related to serialization?

vasily-kirichenko commented on Amqp.V1/Akka.Streams.Amqp.V1/Util/AttributesExtension.cs:
```diff
@@ -0,0 +1,13 @@
+        public static Decider GetDeciderOrDefault(this Attributes attributes)
+        {
+            var attr = attributes.GetAttribute<ActorAttributes.SupervisionStrategy>(null);
+            return attr != null ? attr.Decider : Deciders.StoppingDecider;
```
`return attr?.attr.Decider ?? Deciders.StoppingDecider`?

AndreSteenbergen committed:
Changes made after review

AndreSteenbergen committed:
Added support for ActiveMQ Artemis queues.

AndreSteenbergen commented:
Tested the latest commit using these brokers:
- TestBroker from amqplite
- ActiveMQ 5.15.4
- ActiveMQ Artemis 2.6.2

All seem to work now.

AndreSteenbergen committed:
Merge remote-tracking branch 'upstream/dev' into dev

AndreSteenbergen committed:
Made caal sync again in test.

AndreSteenbergen committed:
Updates interface, so no switch statement is needed to perform actions
Added an event handler to the subscription to raise event when topic eof reached

AndreSteenbergen committed:
Revert "Updates interface, so no switch statement is needed to perform actions"

This reverts commit 44e63db4f1bf9760bfe2ac3da852c21a6a2289a3.

AndreSteenbergen commented:
@marcpiechura What do you think after the updates?

Aaronontheweb commented:
@AndreSteenbergen @marcpiechura @vasily-kirichenko any word on this? I'd like to push out the Akka.Streams.Kafka package but if this is ready for merge, we can do both at the same time.

vasily-kirichenko commented:
Do the two tests cover all this PR functionality?

AndreSteenbergen commented:
Sorry was away on vacation; I think all functionality is captured. With amqp 1 the server implements logic, instead of the client. I couldn't think of more tests besides receiving what was send.

AndreSteenbergen committed:
Updated NuGet packget

Horusiath commented:
@marcpiechura what is the status of this PR? Could we merge it?

Danthar commented:
Im merging this :P

Aaronontheweb commented:
@Danthar thank you 

https://www.youtube.com/watch?v=SxMPTEeoShU

Let's prep a new release here, along with the updated Kafka stuff if that's ready to go
